libraries/AP_TemperatureSensor/AP_TemperatureSensor_MLX90614.cpp

```diff
@@ -0,0 +1,65 @@
+#include "AP_TemperatureSensor_MLX90614.h"
+
+#if AP_TEMPERATURE_SENSOR_MLX90614_ENABLED
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_HAL/I2CDevice.h>
+#include <AP_Math/AP_Math.h>
+
+
+extern const AP_HAL::HAL &hal;
+
+#define MLX90614_I2CDEFAULTADDR 0x5A    // Device default slave address 
+#define MLX90614_BROADCASTADDR  0       // Device broadcast slave address 
+
+//  RAM addresses
+#define MLX90614_RAWIR1         0x04    //  RAM reg - Raw temperature, source #1 
+#define MLX90614_RAWIR2         0x05    //  RAM reg - Raw temperature, source #2 
+#define MLX90614_TA             0x06    //  RAM reg - Linearized temperature, ambient 
+#define MLX90614_TOBJ1          0x07    //  RAM reg - Linearized temperature, source #1 
+#define MLX90614_TOBJ2          0x08    //  RAM reg - Linearized temperature, source #2 
+
+void AP_TemperatureSensor_MLX90614::init() {
+
+    if ((_params.bus_address != MLX90614_I2CDEFAULTADDR) ) {
+        _params.bus_address.set(MLX90614_I2CDEFAULTADDR);
+    }
+    
+    _dev = std::move(hal.i2c_mgr->get_device(_params.bus, _params.bus_address));
+    if (!_dev) {
+        return;
+    }
+
+    WITH_SEMAPHORE(_dev->get_semaphore());
+
+    _dev->register_periodic_callback(50 * AP_USEC_PER_MSEC,
+                                     FUNCTOR_BIND_MEMBER(&AP_TemperatureSensor_MLX90614::_timer, void));
+}
+
+
+void AP_TemperatureSensor_MLX90614::_timer() {
+
+     _crude_value = read_data(MLX90614_TA);
+    
+    if(_crude_value > 0) {
+
+        WITH_SEMAPHORE(_dev->get_semaphore());
+        
+        // temp * 0.02 - 273.15 = degrees, temp * 0.02 is temperature in kelvin
+        const float tmp = KELVIN_TO_C(_crude_value) * 0.02;
+        set_temperature(tmp);
+    }
+}
+
+
+
+uint16_t AP_TemperatureSensor_MLX90614::read_data(uint8_t cmd) {
```

Comment:
You could probably just fold this back into the calling function.

The "0-on-failure" return is probably OK log-wise but a little odd in that you're using an invalid integer value to convey failure.

I'm not all that fussed but other people at DevCall might not like it.

Reply:
> if I return -1 in case of failure that will be fine ?

I'm fine with it.  The usual pattern would be to return bool and take a reference to a value to be filled in on success.

But again I'll note that there's just not much *in* the function, and it's only used on one place :-)

Reply:
Got the point, I am changing this like this which would be quite better  -
`
void AP_TemperatureSensor_MLX90614::_timer() {

    if(read_data(MLX90614_TA, _crude_value)) {

        WITH_SEMAPHORE(_dev->get_semaphore());
        
        // temp * 0.02 - 273.15 = degrees, temp * 0.02 is temperature in kelvin
        const float tmp = KELVIN_TO_C(_crude_value) * 0.02;
        set_temperature(tmp);
    }
}



bool AP_TemperatureSensor_MLX90614::read_data(uint8_t cmd, uint16_t &crude_value) {
    
    if (!_dev->transfer(&cmd, 1, val, 3))
    {
        return false;
    }
    crude_value =  UINT16_VALUE(val[1],val[0]);
    return true;
}`